mypy/semanal.py

```diff
@@ -2944,7 +2944,7 @@ def analyze_simple_literal_type(
         """Return builtins.int if rvalue is an int literal, etc.
 
         If this is a 'Final' context, we return "Literal[...]" instead."""
-        if self.options.semantic_analysis_only or self.function_stack:
+        if self.options.semantic_analysis_only or self.function_stack and not do_bools:
```

Comment:
does that mean it only works on literals? what about
```py
foo: int

def bar(value=foo):
    ...
```

Reply:
It's only ever worked for literals, refs are harder